### PR TITLE
Get `filesize()` if `file_exists()`

### LINKED ISSUE #28717: `files_versions` expiration job trying to read filesize of (already) deleted files
- NC 21.0.4

After set `'versions_retention_obligation' => 'auto, 7',` observe in the logs (hundreds of lines for different files) in a daily baisis:

```
{
   "reqId":"eTxII8UuevYHseynuf6M",
   "level":3,
   "time":"2021-09-04T21:14:17+00:00",
   "remoteAddr":"",
   "user":"--",
   "app":"PHP",
   "method":"",
   "url":"--",
   "message":{
      "Exception":"Error",
      "Message":"filesize(): stat failed for /data/user/files_versions/folder/file.txt.v1628691872 at /lib/private/Files/Storage/Local.php#227",
      "Code":0,
      "Trace":[
         {
            "function":"onError",
            "class":"OC\\Log\\ErrorHandler",
            "type":"::",
            "args":[
               2,
               "filesize(): stat failed for /data/user/files_versions/folder/file.txt.v1628691872",
               "/lib/private/Files/Storage/Local.php",
               227,
               {
                  "path":"files_versions/folder/file.txt.v1628691872",
                  "fullPath":"/data/user/files_versions/folder/file.txt.v1628691872"
               }
            ]
         },
         {
            "file":"/lib/private/Files/Storage/Local.php",
            "line":227,
            "function":"filesize",
            "args":[
               "/data/user/files_versions/folder/file.txt.v1628691872"
            ]
         },
         {
            "file":"/lib/private/Files/Storage/Wrapper/Wrapper.php",
            "line":157,
            "function":"filesize",
            "class":"OC\\Files\\Storage\\Local",
            "type":"->",
            "args":[
               "files_versions/folder/file.txt.v1628691872"
            ]
         },
         {
            "file":"/lib/private/Files/Storage/Wrapper/Wrapper.php",
            "line":157,
            "function":"filesize",
            "class":"OC\\Files\\Storage\\Wrapper\\Wrapper",
            "type":"->",
            "args":[
               "files_versions/folder/file.txt.v1628691872"
            ]
         },
         {
            "file":"/lib/private/Files/View.php",
            "line":1172,
            "function":"filesize",
            "class":"OC\\Files\\Storage\\Wrapper\\Wrapper",
            "type":"->",
            "args":[
               "files_versions/folder/file.txt.v1628691872"
            ]
         },
         {
            "file":"/lib/private/Files/View.php",
            "line":415,
            "function":"basicOperation",
            "class":"OC\\Files\\View",
            "type":"->",
            "args":[
               "filesize",
               "files_versions/folder/file.txt.v1628691872"
            ]
         },
         {
            "file":"/apps/files_versions/lib/Storage.php",
            "line":587,
            "function":"filesize",
            "class":"OC\\Files\\View",
            "type":"->",
            "args":[
               "files_versions///folder/file.txt.v1628691872"
            ]
         },
         {
            "file":"/apps/files_versions/lib/Storage.php",
            "line":498,
            "function":"getAllVersions",
            "class":"OCA\\Files_Versions\\Storage",
            "type":"::",
            "args":[
               "user"
            ]
         },
         {
            "file":"/apps/files_versions/lib/BackgroundJob/ExpireVersions.php",
            "line":66,
            "function":"expireOlderThanMaxForUser",
            "class":"OCA\\Files_Versions\\Storage",
            "type":"::",
            "args":[
               "user"
            ]
         },
         {
            "file":"/lib/private/User/Manager.php",
            "line":646,
            "function":"OCA\\Files_Versions\\BackgroundJob\\{closure}",
            "class":"OCA\\Files_Versions\\BackgroundJob\\ExpireVersions",
            "type":"->",
            "args":[
               "*** sensitive parameters replaced ***"
            ]
         },
         {
            "file":"/apps/files_versions/lib/BackgroundJob/ExpireVersions.php",
            "line":67,
            "function":"callForSeenUsers",
            "class":"OC\\User\\Manager",
            "type":"->",
            "args":[
               {
                  "__class__":"Closure"
               }
            ]
         },
         {
            "file":"/lib/private/BackgroundJob/Job.php",
            "line":52,
            "function":"run",
            "class":"OCA\\Files_Versions\\BackgroundJob\\ExpireVersions",
            "type":"->",
            "args":[
               null
            ]
         },
         {
            "file":"/lib/private/BackgroundJob/TimedJob.php",
            "line":59,
            "function":"execute",
            "class":"OC\\BackgroundJob\\Job",
            "type":"->",
            "args":[
               {
                  "__class__":"OC\\BackgroundJob\\JobList"
               },
               {
                  "__class__":"OC\\Log"
               }
            ]
         },
         {
            "file":"/cron.php",
            "line":128,
            "function":"execute",
            "class":"OC\\BackgroundJob\\TimedJob",
            "type":"->",
            "args":[
               {
                  "__class__":"OC\\BackgroundJob\\JobList"
               },
               {
                  "__class__":"OC\\Log"
               }
            ]
         }
      ],
      "File":"/lib/private/Log/ErrorHandler.php",
      "Line":92,
      "CustomMessage":"--"
   },
   "userAgent":"--",
   "version":"21.0.4.1",
   "id":"6133e1aab3887"
}
```

### PR DESCRIPTION
**Pull request by solracsf**
Should make sense.

Fix https://github.com/nextcloud/server/issues/28717

### CODE REVIEW

**Review by blizzz — APPROVED** [review]

**Comment by juliushaertl** [comment]
Seems a bit more like hiding the symptoms, do you have any idea why the files are present in the getDirectoryContent call but not when checking the filesize afterwards? 

https://github.com/nextcloud/server/blob/5f101d910b8cb0c8cd2c36bb896d0e2acacfe2ae/apps/files_versions/lib/Storage.php#L590

**Comment by acsfer** [comment]
There is no reason as storage is accessed only by nextcloud (no external access).

**Comment by acsfer** [comment]
> do you have any idea why the files are present in the getDirectoryContent call but not when checking the filesize afterwards?

@juliushaertl i've possibly found the root cause; files are on a NFSv4 link, not a very fast one.

2 servers are running in front of it, production with Apache + PHP, and a spare, with PHP + MySQL + Redis. NFW is mounted on both. Cronjobs are on the second server; so this might be the cause of this race condition?

Nevertheless, this fixes those errors.

**Review by juliusknorr — APPROVED** [review]

File: lib/private/Files/Storage/Local.php [hunk 1]
```diff
@@ -216,7 +216,7 @@ public function filetype($path) {
 	}
 
 	public function filesize($path) {
-		if ($this->is_dir($path)) {
+		if ($this->is_dir($path) || !$this->file_exists($path)) {
```
Comment (icewind1991) [inline lib/private/Files/Storage/Local.php:219]: `if (!$this->is_file($path)) {`

**Review by icewind1991 — APPROVED** [review]

**Pull request merged**